lib/autoproj/ops/configuration.rb

```diff
@@ -1,9 +1,70 @@
+require "rgl/adjacency"
+require "rgl/dot"
+require "rgl/topsort"
+
 module Autoproj
     module Ops
         #--
         # NOTE: indentation is wrong to let git track the history properly
         #+++
 
+        # PackageSetHierachy be used to build the hierarchy of package set
+        # imports, as directed acyclic graph (DAG)
+        # so that they can be (topologically) sorted according to their
+        # dependencies
+        class PackageSetHierarchy
+            attr_reader :imports_of, :dag
+
+            def initialize(package_sets, root_pkg_set)
+                @dag = RGL::DirectedAdjacencyGraph.new
+
+                @imports_of = Hash.new
+                package_sets.each do |p|
+                    dependencies = PackageSet.resolve_imports(p)
+                    dependencies.each do |dep|
+                        @dag.add_edge dep, p
+                    end
+                    @imports_of[p] = dependencies
+                end
+
+                @dag.add_vertex root_pkg_set
+                import_order = root_pkg_set.imports.to_a
+                import_order.each_with_index do |p, index|
+                    if index + 1 < import_order.size
+                        @dag.add_edge p, import_order[index + 1]
+                        @dag.add_edge p, root_pkg_set
+                    end
+                end
+
+                unless @dag.acyclic?
+                    raise "PackageSet imports contain cycles: #{@dag.cycles}"
+                end
+            end
+
+            def validate_import_order(pkg_set)
+                imports = pkg_set.imports.to_a
+                until imports.empty?
+                    pkg_set = imports.shift
+                    imports.each do |p|
+                        if @imports_of[pkg_set].include?(p)
+                            raise "PackageSet '#{pkg_set.name}' depends on '#{p.name}', "\
+                                  "but is listed before it in the manifest"
+                        end
+                    end
+                end
+            end
+
```

Comment:
The acyclic? test already takes care of what this was doing ... Maybe remove ? (which would get rid of @imports_of)